File: cpp/train.cpp
#include <chrono>
#include "../model.cpp"
using namespace std;

int main(int argc, char *argv[]){
	PyTrainer* trainer = new PyTrainer(8, 1, 4, 0.1);
	trainer->add_textfile("../../../dataset/japanese.txt");
	trainer->compile();
	cout << "#words: " << trainer->get_num_words() << endl;
	cout << "#characters: " << trainer->get_num_characters() << endl;

	int itr = 1;
    auto start = chrono::system_clock::now();
	while(true){
		trainer->perform_mcmc();
		if(itr % 1000 == 0){
		    auto diff = chrono::system_clock::now() - start;
			cout << "itr: " << itr << " - log likelihood: " << trainer->compute_joint_log_likelihood() << " - MCMC acceptance: " << trainer->get_acceptance_rate() << " - precision: " << trainer->compute_mean_precision(0.99, 20) << " - time (s): " << chrono::duration_cast<std::chrono::seconds>(diff).count()
 << endl;
		    start = chrono::system_clock::now();
		}
		if(itr % 10000 == 0){
			trainer->save("glm2.model");
		}
		itr++;
	}
	delete trainer;
}

File: python/train.py
# coding: utf-8
from __future__ import division
from __future__ import print_function
import argparse, os
import model

def main(args):
	trainer = model.trainer(args.coverage, args.cmax, args.tmax, args.sigma)

	# テキストファイルの追加
	assert os.path.exists(args.input)
	if args.input.endswith(".txt"):
		print("loading", args.input)
		trainer.add_textfile(args.input)
	else:
		files = os.listdir(args.input)
		for filename in files:
			if filename.endswith(".txt"):
				print("loading", filename)
				trainer.add_textfile(args.input + "/" + filename)

	trainer.compile()
	print("#words: %d" % trainer.get_num_words())
	print("#characters: %d" % trainer.get_num_characters())

	itr = 1
	while True:
		trainer.perform_mcmc()
		if itr % 1000 == 0:
			print("itr: {} - log likelihood: {} - MCMC acceptance: {} - precision: {}".format(itr, trainer.compute_joint_log_likelihood(), trainer.get_acceptance_rate(), trainer.compute_mean_precision(args.theta, args.max_word_length)))

		if itr % 10000 == 0:
			trainer.save(args.model_filename)
		itr += 1

if __name__ == "__main__":
	parser = argparse.ArgumentParser()
	parser.add_argument("-i", "--input", type=str, default=None, help="訓練用のテキストファイルが入っているディレクトリ.")
	parser.add_argument("-m", "--model-filename", type=str, default="out/model.glm", help="モデルを保存するファイルへのパス.")
	parser.add_argument("-cover", "--coverage", type=int, default=8, help="後ろの何文字から素性ベクトルを作るか.")
	parser.add_argument("-cmax", "--cmax", type=int, default=1, help="後ろの何文字のIDを素性にするか.")
	parser.add_argument("-tmax", "--tmax", type=int, default=4, help="後ろの何文字の文字種を素性にするか.")
	parser.add_argument("-sigma", "--sigma", type=float, default=0.2, help="ランダムウォーク幅.")
	parser.add_argument("-theta", "--theta", type=float, default=0.99, help="単語長の予測の閾値.")
	parser.add_argument("-k", "--max-word-length", type=int, default=15, help="単語の最大長.")
	main(parser.parse_args())